wandb/sdk/wandb_init.py

```diff
@@ -579,6 +578,12 @@ def _make_run_disabled(self) -> Run:
             "_finish",
         ):
             setattr(drun, symbol, lambda *_, **__: None)  # type: ignore
+
+        class MockArtifact:
+            def __getattr__(self, _):
+                return lambda *args, **kwargs: None
+
+        drun.log_artifact = lambda *_, **__: MockArtifact()
```

Comment:
**question:** Was wondering, is `disabled` mode only meant for use in tests?
- If yes, we should probably make proper use of the (built-in) `Mock` class -- though I wouldn't immediately block on this yet since doing it safely could be a larger refactor
- (If no -- ignore this.  Doesn't seem like a great idea to be passing around `Mock` objects in actual runtime code)

Reply:
yeah asking CS they said its mainly used for testing their code without wandb failing anything.
I believe sdk team is aware there needs to be some refactoring: https://weightsandbiases.slack.com/archives/C0108SLEP6K/p1729282166167099

Reply:
There's a way to solve this with a descriptor -- which is usually overcomplicating things but I think makes sense here, esp since it should be a no-op.

(Can sync about it offline in a bit + will also follow up soon with inline example/suggestion)

Reply:
**suggestion:** Following up on the descriptor suggestion -- 

This should allow arbitrarily chained attributes access + methods calls (all no-ops), e.g. `run.log_artifact(artifact).wait().add_file(...)` or similar.  (Though you probably want to verify or test this in case I missed something)

```suggestion

        class _ChainableNoOp:
            """An object that allows chaining arbitrary attributes and method calls."""
            def __getattr__(self, _: str) -> Self:
                return self
                
            def __call__(self, *_: Any, **__: Any) -> Self:
                return self
                
        class _ChainableNoOpField:
            def __init__(self) -> None:
                self._value = None

            def __set__(self, instance: Any, value: Any) -> None:
                self._value = value

            def __get__(self, instance: Any, owner: type) -> Any:
                return _ChainableNoOp() if (self._value is None) else self._value

        drun.log_artifact = _ChainableNoOpField()
```